### ISSUE TITLE
New parameter

### ISSUE THREAD

**Issue by PB12203006** [Stevo15025/tsp_gpu#10]
Our parameters are not well chosen so I changed the method on adjusting sample space and higher loss function acceptance probability.

I read the article and think it dose not suit our tsp problem, it's because in physics, the energy is a function of T, they are correlated, and all the evolution of parameters on T is based on this relation, so terms like exp(-E/T) works well.

However, for tsp, the T is just a number, and it's totally independent of all things.

Thus, in physics questions, the range of exp(-E/T) is restricted, but look at out dsp problem, in acceptance, we originally use exp(-delta/T), but what's delta? It depends on the data we import, if the locations are very narrow, delta might be very small, then all the exp(-delta/T)  are near 1, it means we may accept all changes without good filter.

And for our data, the locations are sparse, and the delta is very big, that makes the acceptance probability goes near 0, meaning we can hardly accept any higher loss change.

So, to get a proper acceptance probability, I changed the calculation on acceptance, I use

        p=(original_dist / proposal_dist)*T/2

and set T to decrease slowly from 1 to 0, while iteration, T=T*beta, beta<1

to be the acceptance probability. So at the beginning we can accept slight increase of loss function, and when T goes down, that gradually becomes forbidden.

The evolution of sample space is based on the same principle, I just set it to N*T

Because exp(...) is hard to control, for example, for N=100K, we have to let exp(...) decrease to 1/100K, it's hard to set.

And the last change I made is, I cancelled the inner iteration in threads in swap function, I think iter in gpu is slow, we don't have to do so. But while writing now I think iter is not as bad as I thought before, maybe you can add it back if you like...

**Comment by PB12203006** [comment]
Plus, I use two T now, because the insertion sample space is restricted to around 9000, It cannot decrease simultaneously with swapping's sample space, So I separate the two temperature, when swapping temp goes down so that its sample space decrease to insertion's, the insertion's space begin to decrease.

**Comment by Stevo15025** [comment]
> However, for tsp, the T is just a number, and it's totally independent of all things.

I'm not sure if I agree that T has any real "objective" meaning when used in either physics or statistics. Either way the whole principle of SA is that it has a reasonable acceptance/rejection step for the Metropolis-Hasting's algorithm so we can nicely search the over the graph of some function for the minimum.

> So, to get a proper acceptance probability, I changed the calculation on acceptance

For the record, changing the acceptance/rejection step away from it's standard definition makes this no longer simulated annealing. This also means we are starting to throw out some very good theory. 

I definitely agree that scale and sparsity is an issue we should try to fix but I don't think there is a need to throw the baby out with the bathwater. Why don't we just scale the loss function? 

```
loss_scaled = (original_loss - proposed_loss) / original_loss
```

then we would just plug this into the acceptance/rejection step

```
accept = e^(- (loss_scaled / T))
```

>  I think iter in gpu is slow, we don't have to do so.

I believe it looks slower because we are only tracking outer iterations in the C code, so it looks like the inner loop slows things down quite a lot. Really though, it just allows us to run more samples at a given temperature. It's just exploring a larger part of the space before it gives up and moves to a lower temperature.

**Comment by PB12203006** [comment]
Yes, your argument is really reasonable.

First, iter in gpu is ok. It's not a bad idea. I apologize.

As for the evolution rule, also, you're right, I made a mistake, exp is good. But be careful on the range of T.  It varies with different city numbers. because at the end of annealing, the sample space should decrease to 1, that indicates at the end

    N*exp(-parameter_1/T_end)=1

but at the beginning, it should be near N, and more importantly, we sample uniformly, that means even the space is N, the average range we sample is N/2, so I think we have to keep the space large for longer  time, in order to get enough long-range swap. So parameter_1 should be a small number, or use a large T_start. And then to control the running time we may have to adjust the evolution speed, like 

    iter+=2, or 0.5

As for acceptance, also, we need a parameter_2 to adjust, and put a factor above, like

    0.5*exp(-parameter_2/T)

And more importantly, separating T_0 and T_1 is very very necessary. Because of the restriction of sample space in insertion. At beginning the sample space in swapping is N while in insertion it's only 1000, so if they evolve by the same method, when swapping ranges N/1000, the insertion goes down to 1. That's obviously not good, so my suggestion is, let insertion-sample-space fix to 1000, wait until the swap-sample-space decrease to 1000, then it begins to decrease synchronizing with swap's. 

I've tried, adding insertion can improve the effect a lot, it's a great deal. As I illustrated, each method is good at it's typical pattern and using them alternatively can improve the performance a lot, even if the space of insertion is restricted to 1000. And space 1000 is stable and implementable. We have no risk implementing insertion restricted in 1000 now.

**Comment by PB12203006** [comment]
And I strongly suggest that adjust the random number range to (0.9, 1) in swap might be a better idea. Because uniformly sample is not optimized, for example, what if we need to swap i and i+1000 and the space is 1100 by then, but for many times the program always samples ranges like 800, 600, 200, 60, 500.... As you know, it's hard to sample edge if we sample uniformly. 

And it's very easy to implement: assume x is a random number in (0,1) then just transform

    y=0.9+x/10

And y is in (0.9,1)

**Comment by PB12203006** [comment]
Serious numerical problem:

If we evolve T by 

    T=T_start/log(iter)

Then, here's my simple calculation:

If

    sample space= N*exp(parameter/T)

Where N=100K
First, at the beginning, the factor should be near 1: 

    exp(-0.1)=0.9

So, 

    parameter/T_start=0.1        (1)

and, in the end, the space should decrease to 1, that means

    exp(-parameter/T_end)=10^-5

So,

    parameter/T_end=-11.5      (2)

let (2)/(1)

    T_start=115*T_end

Now here's a serious problem:

    iter_end=exp(T_start/T_end)=exp(115)

It's far far longer than unsigned long long int, we cannot define a variable iter has such long range!

So by now my suggestion is, keep the term exp(...) in the sample space and acceptance, but outside the kernel, T itself evolve by

    T=T*beta

This method is generally used by many paper, I think there's no problem. Or if you can solve this overflow problem, log(iter) is ok.

**Comment by PB12203006** [comment]
And I think things in the kernel are already done, I tried several methods, but shifting many elements of an array is time-consuming so I think doing long-range insertion is meaningless, this method can only be a supportive method when the T goes down. And it's stable and safe when we restrict the range within 1024.

**Comment by Stevo15025** [comment]
Let me have one final look over everything and I'll merge this branch to master

**Comment by PB12203006** [comment]
One correction: modify the acceptance to

    exp(-2000*quotient/T)

then for 10% increase, at the beginning acceptance is exp(-200/120)=18.89% and decay to nearly zero at the end.

The original version is too large,
 
   exp(-40*0.1/120)=96%

even 

    exp(-40*0.1/30)=87%

That's a ridiculous number, I made a mistake yesterday.

And no matter what size the route is, the T_start is fixed to 120, and the sample space at the beginning is near 1, and we should adjust the T_end corresponding to city number N

    exp(-12/T_end)=1/N

Then 

    T_end=12/log(N)

**Comment by PB12203006** [comment]
Let me explain the principle of selection of parameters:

First of all, T_start

    sample_space=exp(-A/T)*N

At the beginning, for example if I select

    sample_space_start=0.9N

Then 

    exp(-A/T_start)=0.9

    T_start=-A/log(0.9)

What really matters is the quotient, So we can select A whatever we like. if I set A to 12, then T_start=120, it's just an example, you can also set

    exp(-A/T_start)=0.99

then just set a lovely A and calculate the corresponding T_start

And, at the end, the sample space should be 1, so

    exp(-A/T_end)*N=1

That means 
  
    T_end=A/log(N)

**It does not matter that T_end depends on N, because we can always automatically derivate T_end, so it's friendly, just write the host code as:**

    while( T > 12/log(N) )
        {  do annealing  }

Actually it's **inevitable** because we can never find a fixed T to solve any tsp problem, when N goes large, correspondingly the range of T expands, it's obvious, because the problem becomes more complicated, thus we need calculate more. But by this condition (T>12/log(N)) we make the selection of T automatic, I think it's tricky but friendly. And again, T_start is independent of N, just fix a number satisfies what I said above.

So, in conclusion, I set

    T_end=1 for N=100K  and   sample_space_start=0.9N

that's why I get the parameters

    A=12, T_start=120, T_end=1

If you want T starts from 1, and sample_space_start=0.99N

    A=-log(0.99)=0.01    T_end=0.01/log(N)

And for acceptance, no strict restriction, first remember that I substracted 1 from quotient, so quotient=0.1 implies increasing 10% distance
 just guarantee that

    acceptance_start(10%)=(-quotient*B/T_start)~=10%-20%

and

    acceptance_near_end(10%)=(-quotient*B/(T_start*0.3))~=0

And I think it's ok.

For my 12_120 parameters, B=40 is good, you can tryout your preferred parameter easily.

**Comment by Stevo15025** [comment]
Added in  5e40803
